docs: FE Feedback fixes (#210)
canonical/spark-k8s-bundle: docs/explanation/component-overview.md

(explanation-component-overview)=
# Components overview

Charmed Apache Spark is composed of foundational software artifacts and a set of Juju operators (charms) that together provide a fully managed Apache Spark platform on Kubernetes. All charms are available individually on [Charmhub](https://charmhub.io/) and can also be deployed together via the [Charmed Apache Spark bundle](https://charmhub.io/spark-k8s-bundle) or [Terraform modules](https://github.com/canonical/spark-k8s-bundle/tree/main/releases/3.4/terraform).

## Software artifacts

### spark8t

> [spark8t](https://github.com/canonical/spark-k8s-toolkit-py) is a Python library that extends Apache Spark with tooling to manage Spark jobs and service accounts with hierarchical configuration. It is the foundation shared by both the `spark-client` snap and the Juju charms.

Reviewer: ```suggestion
[spark8t](https://github.com/canonical/spark-k8s-toolkit-py) is a Python library that extends Apache Spark with tooling to manage Spark jobs and service accounts with hierarchical configuration. It is the foundation shared by both the `spark-client` snap, the OCI images and the Juju charms.
```

### Charmed Apache Spark Rock

The [Charmed Apache Spark Rock](https://github.com/canonical/charmed-spark-rock/pkgs/container/charmed-spark) is an OCI-compliant container image that bundles Apache Spark binaries together with Canonical tooling. It is used as the base image for Spark driver and executor pods on Kubernetes, and as the foundation for the `spark-client` snap.

### spark-client snap

The [spark-client snap](https://snapcraft.io/spark-client) provides CLI tools for working with Charmed Apache Spark from a workstation or edge node. It communicates with the Kubernetes API to submit jobs and manage service accounts — it does not connect to any Juju charm directly.

| Command | Description |
|---|---|
| `spark-client.spark-submit` | Submit Spark applications to a Kubernetes cluster |
| `spark-client.pyspark` | Start an interactive PySpark shell |
| `spark-client.service-account-registry` | Create, configure, and manage Spark service accounts |
| `spark-client.beeline` | JDBC client for connecting to Apache Kyuubi endpoints |
| `spark-client.import-certificate` | Import TLS certificates for encrypted Kyuubi connections |

## Juju operators (charms)

### Core integrators

> The following charms form the foundation of any Charmed Apache Spark deployment, connecting Spark service accounts to object storage:

Reviewer: ```suggestion
The following charms form the foundation of any Charmed Apache Spark deployment, connecting Spark service accounts to external services, e.g. object storage or Canonical Observability Stack deployments:
```

| Charm | Description |
|---|---|

> | [`spark-integration-hub-k8s`](https://charmhub.io/spark-integration-hub-k8s) | Central hub that manages Spark service account configurations and writes them into Kubernetes Secrets. Connects to object storage integrators, Kyuubi, and optionally to the COS observability bridge. |

Reviewer: ```suggestion
| [`spark-integration-hub-k8s`](https://charmhub.io/spark-integration-hub-k8s) | Central hub that manages Spark service account configurations and writes them into Kubernetes Secrets. It allows high-level configuration of Spark properties and seamless integration with external services, such as object storage backends and COS deployments. |
```

| [`s3-integrator`](https://charmhub.io/s3-integrator) | Supplies S3-compatible object storage credentials (endpoint, bucket, access key) to the Integration Hub and History Server. Supports MinIO, AWS S3, and any S3-compatible backend. |
| [`azure-storage-integrator`](https://charmhub.io/azure-storage-integrator) | Alternative to `s3-integrator` for deployments using Azure Blob Storage. |

> ### History Server

Reviewer: **suggestion** I would probably put this in the `core` components. Traefik is a rather general service, and it could be deployed one ingress for multiple applications. In some deployments, one could even reuse the one in COS. And this would only leave `spark-history-server-k8s` alone, so better to factor this in the `core` part IMHO

Author: Good idea. Fixed in https://github.com/canonical/spark-k8s-bundle/pull/210/commits/7c6b7ab43c0388739e915f1085a9c3b88e67bc7e

            direction LR
            s3int["s3-integrator"]
            azint["azure-storage-integrator"]
        end

        subgraph hist-grp["History Server"]
            direction LR
            hs["spark-history-server-k8s"]
            traefik["traefik-k8s"]
        end

        subgraph kyu-grp["Apache Kyuubi"]
            direction TB
            kyuubi["kyuubi-k8s"]
            authdb["postgresql-k8s<br>(auth-db)"]
            metadb["postgresql-k8s<br>(metastore)"]
            zk["zookeeper-k8s"]
            tls["self-signed-certificates"]
            di["data-integrator"]
        end

        subgraph cos-bridge["COS Bridge"]
            direction LR
            pgw["prometheus-pushgateway-k8s"]
            scrape["prometheus-scrape-config-k8s"]
            agent["grafana-agent-k8s"]
            cosconf["cos-configuration-k8s"]
        end
    end

    subgraph cos-model["COS Juju Model (cos-lite)"]
        direction LR
        prom["Prometheus"]
        graf["Grafana"]
        loki_n["Loki"]
    end

    client -->|"K8s API<br>(spark-submit / pyspark)"| hub
    client -->|"JDBC"| kyuubi
    backend <-->|credentials| s3int & azint

>     s3int & azint -->|s3-credentials| hub

Reviewer: **todo** This is a bit confusing as also for the rendering it seems that azure storage integrator has an `s3-credentials` relation. I believe we should not have `s3int` and `azint` together here, but rather have two lines separated one for `s3-credentials` and the other for `azure-storage-credentials`

Author: Fixed in https://github.com/canonical/spark-k8s-bundle/pull/210/commits/7c6b7ab43c0388739e915f1085a9c3b88e67bc7e

    s3int & azint -->|s3-credentials| hs
    hub -->|service-account| kyuubi
    hub -->|cos| pgw
    traefik -->|ingress| hs

>     kyuubi --- authdb

Reviewer: **todo** maybe best to also add the name of the relation here, something like `postgresql_client`. Otherwise these lines do not have any label, like the others

Author: I've streamlined a bit the labels (with the help of AI) for consistent naming practices based on endpoints. See commit https://github.com/canonical/spark-k8s-bundle/pull/210/commits/7c6b7ab43c0388739e915f1085a9c3b88e67bc7e

Current labels in the diagram and their source endpoints:

Label | Direction | Charm endpoint
-- | -- | --
s3-credentials | s3int → hub, s3int → hs | spark-integration-hub-k8s requires s3-credentials
azure-storage-credentials | azint → hub, azint → hs | spark-integration-hub-k8s requires azure-storage-credentials
spark-service-account | hub → kyuubi | spark-integration-hub-k8s provides spark-service-account
cos | hub → pgw | spark-integration-hub-k8s requires cos
ingress | traefik → hs | spark-history-server-k8s requires ingress
auth-db | kyuubi → authdb | kyuubi-k8s requires auth-db
metastore-db | kyuubi → metadb | kyuubi-k8s requires metastore-db
zookeeper | kyuubi → zk | kyuubi-k8s requires zookeeper
certificates | kyuubi → tls | kyuubi-k8s requires certificates
jdbc | di → kyuubi | kyuubi-k8s provides jdbc

The `credentials` label on the backend↔integrator lines and `K8s API`/`JDBC` on the client lines are descriptive (not Juju relation names), since those connections are not Juju relations.

    kyuubi --- metadb
    kyuubi --- zk
    kyuubi --- tls

>     di -->|get-credentials| kyuubi

Reviewer: ```suggestion
    di -->| kyuubi_client | kyuubi
```
